redhat-appstudio/application-service
# Detect build pipeline to use for component

Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

#### What this MR does

This MR adds functionality to choose build pipeline depending on the component. It uses data stored in component's model to make decision. The algorithm of pipeline detection is the following:
1. Check component's model for language and framework. If they are specified, use available language specific build pipeline.
2. If the language is not detected, then look for dockerfile and use it
3. If nothing above works, return `noop` build as it is unknow how to build the component.

#### How to test

a. _Java Quarkus scenario_
1. Create a sample application
2. Create a component that points to [Quarkus example](https://github.com/devfile-samples/devfile-sample-code-with-quarkus): 
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  name: component-sample
  namespace: test
spec:
  componentName: sample-component-name
  application: application-sample
  source:
    git:
      url: https://github.com/devfile-samples/devfile-sample-code-with-quarkus
  build:
    containerImage: docker.io/username/foobar:dev-code-with-quarkus
```
3. Check that a new PipeplineRun for Java projects is created.

b. _Dockerfile scenario_
1. Create a sample application
2. Create a component that points to a repository ([example](https://github.com/mmorhun-org/devfile-stack-go)) with a simple Go lang server, Dockerfile and Devfile:
```golang
package main
import (
  "fmt"
  "net/http"
)
func main() {
  fmt.Println("Starting server on port 8080")
  http.HandleFunc("/", HelloServer)
  http.ListenAndServe("0.0.0.0:8080", nil)
}
func HelloServer(w http.ResponseWriter, r *http.Request) {
  fmt.Fprintf(w, "Hello, %s!", r.URL.Path[1:])
}
```
```dockerfile
FROM golang
COPY main.go main.go
EXPOSE 8080
ENTRYPOINT go run main.go
```
```yaml
schemaVersion: 2.2.0
metadata:
  name: simple-go
  version: 1.1.0
  displayName: Simple Go lang server
  description: Test project for simple Go lang server
  attributes:
    alpha.dockerimage-port: 8080
components:
  - name: outerloop-build
    image:
      imageName: docker.io/mm4eche/foobar:golang
      dockerfile:
        uri: ./Dockerfile
        buildContext: .
        rootRequired: false
```
3. Check that a new PipeplineRun for Devfile projects is created.
